ocrd/ocrd/workspace.py

```diff
@@ -247,150 +247,270 @@ def _resolve_image_as_pil(self, image_url, coords=None):
         ]
         return Image.fromarray(region_cut)
 
-    def image_from_page(self, page, page_id):
-        """Extract the Page image from the workspace.
-
-        Given a PageType object, `page`, extract its PIL.Image from
+    def image_from_page(self, page, page_id, feature_selector='', feature_filter=''):
+        """Extract a Page image from the workspace.
+        
+        Given a PageType object, ``page``, extract its PIL.Image from
         AlternativeImage if it exists. Otherwise extract the PIL.Image
-        from imageFilename and crop it if a Border exists. Otherwise
-        just return it.
-
-        When cropping, respect any orientation angle annotated for
-        the page (from page-level deskewing) by rotating the
-        cropped image, respectively.
-
+        from imageFilename. Also crop it if a Border exists, and rotate
+        it if an @orientation exists. Otherwise just return it.
+        
+        If ``feature_selector`` and/or ``feature_filter`` is given, then
+        select/filter among imageFilename and all AlternativeImages the
+        last which contains all of the selected but none of the filtered
+        features (i.e. @comments classes), or raise an error.
+        
+        If the chosen image does not have "cropped", but a Border exists,
+        then crop it (unless "cropped" is also being filtered). And if the
+        chosen image does not have "deskewed", but an @orientation exists,
+        then rotate it (unless "deskewed" is also being filtered).
+        
+        Cropping uses a polygon mask (not just the rectangle).
+        
+        (Required and produced features need not be in the same order, so
+        ``feature_selector`` is merely a mask specifying Boolean AND, and
+        ``feature_filter`` is merely a mask specifying Boolean OR.)
+        
         If the resulting page image is larger than the bounding box of
-        `page`, pass down the page's box coordinates with an offset of
-        half the width/height difference.
-
-        Return the extracted image, and the absolute coordinates of
-        the page's bounding box / border (for passing down), and
-        an OcrdExif instance associated with the original image.
+        ``page``, then in the returned bounding box, reduce the offset by
+        half the width/height difference (so consumers being passed this
+        image and offset will still crop relative to the original center).
+        
+        Return a tuple:
+         * the extracted image,
+         * a dictionary with the absolute coordinates of the page's
+           bounding box / border (xywh), angle and the AlternativeImage
```

Comment:
AlternativeImage → AlternativeImage's ???

Reply:
I deliberately avoided the possessive here, because that would carry an ambiguity: is it about the _input_ `AlternativeImage` or the _output_ image that can/should/will be added based on the returned image? The returned value actually is the latter, but with a possessive, it sounds more like the former (to me, anyway). Or should I make that more verbose?